File: backend/app/core/followup.py
from datetime import date

FOLLOWUP_STALE_DAYS = 7
DECISION_STALE_DAYS = 3
TERMINAL_STATUSES = {'Rejected', 'Ghosted'}


def _days_since(d: date, today: date) -> int:
    return (today - d).days


def needs_followup(app, today: date) -> bool:
    if str(app.status) in TERMINAL_STATUSES:
        return False
    if app.followed_up_at is not None:
        return False
    last_activity = app.updated_at.date() if app.updated_at else app.applied_date
    return _days_since(last_activity, today) >= FOLLOWUP_STALE_DAYS


def awaiting_response(app, today: date) -> bool:
    if str(app.status) in TERMINAL_STATUSES:
        return False
    if app.followed_up_at is None:
        return False
    return _days_since(app.followed_up_at, today) < DECISION_STALE_DAYS


def needs_decision(app, today: date) -> bool:
    if str(app.status) in TERMINAL_STATUSES:
        return False
    if app.followed_up_at is None:
        return False
    return _days_since(app.followed_up_at, today) >= DECISION_STALE_DAYS


def classify(app, today: date) -> str:
    """Returns 'needs_followup' | 'awaiting_response' | 'needs_decision' | 'ok'."""
    if needs_decision(app, today):
        return 'needs_decision'
    if awaiting_response(app, today):
        return 'awaiting_response'
    if needs_followup(app, today):
        return 'needs_followup'
    return 'ok'

File: frontend/src/utils/followup.ts
import { JobApplication, ApplicationStatus } from '../types';

export const FOLLOWUP_STALE_DAYS = 7;
export const DECISION_STALE_DAYS = 3;
const TERMINAL = new Set<ApplicationStatus>(['Rejected', 'Ghosted']);

export function daysSince(dateStr: string): number {
  const d = new Date(dateStr);
  const today = new Date();
  d.setHours(0, 0, 0, 0);
  today.setHours(0, 0, 0, 0);
  return Math.floor((today.getTime() - d.getTime()) / 86400000);
}

export type FollowupClass = 'needs_followup' | 'awaiting_response' | 'needs_decision' | 'ok';

export function classifyApp(app: JobApplication): FollowupClass {
  if (TERMINAL.has(app.status)) return 'ok';
  if (app.followedUpAt) {
    const days = daysSince(app.followedUpAt);
    return days >= DECISION_STALE_DAYS ? 'needs_decision' : 'awaiting_response';
  }
  return daysSince(app.updatedAt) >= FOLLOWUP_STALE_DAYS ? 'needs_followup' : 'ok';
}

// Map each non-terminal status to its natural next stage
export const NEXT_STAGE: Partial<Record<ApplicationStatus, ApplicationStatus>> = {
  Shortlisted: 'Applied',
  Applied: 'Replied',
  Replied: 'Phone Screen',
  'Phone Screen': 'Technical Round 1',
  'Technical Round 1': 'Technical Round 2',
  'Technical Round 2': 'Final Round',
  'Final Round': 'Offer',
};
